semgrep/semgrep/join_rule.py

```diff
@@ -0,0 +1,404 @@
+        B = model_map[collection_b_real]
+
+        condition_terms.append(
+            (A, condition.property_a, B, condition.property_b, condition.operator)
+        )
+
+    # Use the rhs of the final condition as the finding to return.
+    # Without this, the return value is non-deterministic.
+    last_condition_model: Type[BaseModel] = condition_terms[-1][2]
+    query = (
+        joined.select(last_condition_model.raw)
+        .distinct()
+        .where(
+            *list(map(lambda terms: evaluate_condition(*terms), condition_terms))  # type: ignore
+        )
+    )
+    logger.debug(query)
+    return query
+
+
+def create_config_map(semgrep_config_strings: List[str]) -> Dict[str, Rule]:
+    config = {}
+    for config_string in semgrep_config_strings:
+        resolved = resolve_config(config_string)
+        # Some code-fu to get single rules
+        config.update(
+            {config_string: list(Config._validate(resolved)[0].values())[0][0]}
+        )
+    return config
+
+
+def rename_metavars_in_place(
+    semgrep_results: List[Dict[str, Any]],
+    refs_lookup: Dict[str, Ref],
+) -> None:
+    for result in semgrep_results:
+        metavars = result.get("extra", {}).get("metavars", {})
+        renamed_metavars = {
+            refs_lookup[result.get("check_id", "")].renames.get(
+                metavar, metavar
+            ): contents
+            for metavar, contents in metavars.items()
+        }
+        result["extra"]["metavars"] = renamed_metavars
+
+
+def create_model_map(
+    semgrep_results: List[Dict[str, Any]]
+) -> Dict[str, Type[BaseModel]]:
+    collections: Dict[str, List[Dict]] = group(
+        semgrep_results, key=lambda item: item.get("check_id")
+    )
+    model_map: Dict[str, Type[BaseModel]] = {}
+    for name, findings in collections.items():
+        metavars = set()
+        for finding in findings:
+            metavars.update(finding.get("extra", {}).get("metavars", {}).keys())
+        model_fields = ["path"] + list(metavars)
+        model_class = model_factory(camel_case(name), model_fields)
+        model_map[name] = model_class
+    return model_map
+
+
+def populate_data(
+    semgrep_results: List[Dict[str, Any]], model_map: Dict[str, Type[BaseModel]]
+) -> None:
+    collections = group(semgrep_results, key=lambda item: item.get("check_id"))
+    for name, findings in collections.items():
+        for finding in findings:
+            model_map[name].create(
+                path=finding.get("path"),
+                raw=json.dumps(finding),
+                **{
+                    metavar: content.get("abstract_content").strip().strip("\"'")
+                    for metavar, content in finding.get("extra", {})
+                    .get("metavars", {})
+                    .items()
+                },
+            )
+
+
+def main(
+    join_rule: Dict[str, Any],
+    targets: List[Path],
+) -> Tuple[List[RuleMatch], List[SemgrepError]]:
```

Comment:
Can we rename from `main` probably `run_join_rules` and add function level comment that this expects a single dictionary join rule and runs all relevant subrules with semgrep and joins as necessary